src/heartratetask/HeartRateTask.cpp

```diff
@@ -110,10 +92,114 @@ void HeartRateTask::StartMeasurement() {
+    state = States::BackgroundMeasuring;
+    StartMeasurement();
+  }
+}
+
+void HeartRateTask::HandleSensorData(int* lastBpm) {
+  int8_t ambient = ppg.Preprocess(heartRateSensor.ReadHrs(), heartRateSensor.ReadAls());
+  int bpm = ppg.HeartRate();
+
+  // If ambient light detected or a reset requested (bpm < 0)
+  if (ambient > 0) {
+    // Reset all DAQ buffers
+    ppg.Reset(true);
+  } else if (bpm < 0) {
+    // Reset all DAQ buffers except HRS buffer
+    ppg.Reset(false);
+    // Set HR to zero and update
+    bpm = 0;
+  }
+
+  if (*lastBpm == 0 && bpm == 0) {
+    controller.Update(Controllers::HeartRateController::States::NotEnoughData, bpm);
+  }
+
+  if (bpm != 0) {
+    *lastBpm = bpm;
+    controller.Update(Controllers::HeartRateController::States::Running, bpm);
+    if (state == States::Measuring || IsContinuosModeActivated()) {
+      return;
+    }
+    if (state == States::BackgroundMeasuring) {
+      state = States::BackgroundWaiting;
+      StartWaiting();
+    }
+  }
+  TickType_t ticksSinceMeasurementStart = xTaskGetTickCount() - measurementStart;
+  if (bpm == 0 && state == States::BackgroundMeasuring && !IsContinuosModeActivated() &&
+      ticksSinceMeasurementStart >= DURATION_UNTIL_BACKGROUND_MEASURMENT_IS_STOPPED) {
+    state = States::BackgroundWaiting;
+    StartWaiting();
+  }
+  if (bpm == 0 && state == States::BackgroundMeasuring &&
+      xTaskGetTickCount() - measurementStart >= DURATION_UNTIL_BACKGROUND_MEASURMENT_IS_STOPPED) {
+    state = States::BackgroundWaiting;
+    StartWaiting();
+  }
+}
+
+TickType_t HeartRateTask::CurrentTaskDelay() {
+  switch (state) {
+    case States::Measuring:
+    case States::BackgroundMeasuring:
+      return ppg.deltaTms;
+    case States::Running:
+      return pdMS_TO_TICKS(100);
+    case States::BackgroundWaiting:
+      return pdMS_TO_TICKS(10000);
+    default:
+      return portMAX_DELAY;
+  }
+}
+
+TickType_t HeartRateTask::GetHeartRateBackgroundMeasurementIntervalInTicks() {
+  switch (settings.GetHeartRateBackgroundMeasurementInterval()) {
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::TenSeconds:
+      return pdMS_TO_TICKS(10 * 1000);
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::ThirtySeconds:
+      return pdMS_TO_TICKS(30 * 1000);
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::OneMinute:
+      return pdMS_TO_TICKS(60 * 1000);
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::FiveMinutes:
+      return pdMS_TO_TICKS(5 * 60 * 1000);
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::TenMinutes:
+      return pdMS_TO_TICKS(10 * 60 * 1000);
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::ThirtyMinutes:
+      return pdMS_TO_TICKS(30 * 60 * 1000);
+    default:
+      return 0;
+  }
+}
```

Comment:
This would only be something like
```
pdMS_TO_TICKS(settings.Get...Interval())
```
if you store the interval values as seconds.

Reply:
Another small idea: would it be beneficial for the image size to store the seconds in a variable and move `pdMS_TO_TICKS` behind the switch? I don't know how the compiler handles that and cannot try it out myself right now.

Reply:
Refactored to only call `pdMS_TO_TICKS` once